From 2ee257682c98ffa05d5624370387d22326fbd217 Mon Sep 17 00:00:00 2001
From: Rajashekhara, Sudhakar <sudhakar.raj@ti.com>
Date: Wed, 2 Dec 2009 05:34:38 +0000
Subject: [PATCH 14/23] davinci: reserve channels/slots only when DSP is present

Original codes from TI Linux Platform Support Package
DaVinci-PSP-SDK-03.20.00.12.tgz http://software-dl.ti.com/dsps/
dsps_public_sw/sdo_sb/targetcontent/psp/DaVinci-PSP-SDK/03_20/index_FDS.html

This patch reserves EDMA channels and slots only when DSP is present
on the SoC. If DSP is not present, all the channels are available
for allocation from ARM.

Signed-off-by: Sekhar Nori <nsekhar@ti.com>
Signed-off-by: Sudhakar Rajashekhara <sudhakar.raj@ti.com>
Integrated-by: Stanley.Miao <stanley.miao@windriver.com>
---
 arch/arm/mach-davinci/devices-da8xx.c |   58 +++++++++++++++++++++++++++++++++
 1 files changed, 58 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-davinci/devices-da8xx.c b/arch/arm/mach-davinci/devices-da8xx.c
index fadd147..591c488 100644
--- a/arch/arm/mach-davinci/devices-da8xx.c
+++ b/arch/arm/mach-davinci/devices-da8xx.c
@@ -112,6 +112,54 @@ static const s8 da850_queue_priority_mapping[][2] = {
 	{-1, -1}
 };
 
+static const s16 da850_dma0_rsv_chans[][2] = {
+	/* (offset, number) */
+	{ 8,  6},
+	{24,  4},
+	{30,  2},
+	{-1, -1}
+};
+
+static const s16 da850_dma0_rsv_slots[][2] = {
+	/* (offset, number) */
+	{ 8,  6},
+	{24,  4},
+	{30, 50},
+	{-1, -1}
+};
+
+static const s16 da850_dma1_rsv_chans[][2] = {
+	/* (offset, number) */
+	{ 0, 28},
+	{30,  2},
+	{-1, -1}
+};
+
+static const s16 da850_dma1_rsv_slots[][2] = {
+	/* (offset, number) */
+	{ 0, 28},
+	{30, 90},
+	{-1, -1}
+};
+
+static const s16 da830_dma_rsv_chans[][2] = {
+	/* (offset, number) */
+	{ 8,  2},
+	{12,  2},
+	{24,  4},
+	{30,  2},
+	{-1, -1}
+};
+
+static const s16 da830_dma_rsv_slots[][2] = {
+	/* (offset, number) */
+	{ 8,  2},
+	{12,  2},
+	{24,  4},
+	{30, 26},
+	{-1, -1}
+};
+
 static struct edma_soc_info da830_edma_info[] = {
 	{
 		.n_channel		= 32,
@@ -252,6 +300,16 @@ static struct platform_device da850_edma_device = {
 int __init da8xx_register_edma(void)
 {
 	struct platform_device *pdev;
+	
+	/* Reserve channels and slots for DSP */
+	if (!cpu_is_davinci_da8xx_arm_only()) {
+		da830_edma_info[0].rsv_chans = da830_dma_rsv_chans;
+		da830_edma_info[0].rsv_slots = da830_dma_rsv_slots;
+		da850_edma_info[0].rsv_chans = da850_dma0_rsv_chans;
+		da850_edma_info[0].rsv_slots = da850_dma0_rsv_slots;
+		da850_edma_info[1].rsv_chans = da850_dma1_rsv_chans;
+		da850_edma_info[1].rsv_slots = da850_dma1_rsv_slots;
+	}
 
 	if (cpu_is_davinci_da830())
 		pdev = &da830_edma_device;
-- 
1.6.5.2

